plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs

```diff
@@ -106,33 +106,71 @@ mkTermToEvaluate ll pv script args = do
     through (liftEither . first DeBruijnError . UPLC.checkScope) appliedT
 
 toMachineParameters :: MajorProtocolVersion -> EvaluationContext -> DefaultMachineParameters
-toMachineParameters _ = machineParameters
+toMachineParameters pv (EvaluationContext lv toSemVar machParsList) =
+    case lookup (toSemVar pv) machParsList of
+        Nothing -> error $ Prelude.concat
+            ["Internal error: ", show lv, " does not support protocol version ", show pv]
+        Just machPars -> machPars
 
 {-| An opaque type that contains all the static parameters that the evaluator needs to evaluate a
-script.  This is so that they can be computed once and cached, rather than being recomputed on every
+script. This is so that they can be computed once and cached, rather than being recomputed on every
 evaluation.
+
+Different protocol versions may require different bundles of machine parameters, which allows us for
+example to tweak the shape of the costing function of a builtin, so that the builtin costs less.
+Currently this means that we have to create multiple 'DefaultMachineParameters' per language
+version, which we put into a cache (represented by an association list) in order to avoid costly
+recomputation of machine parameters.
+
+In order to get the appropriate 'DefaultMachineParameters' at validation time we look it up in the
+cache using a semantics variant as a key. We compute the semantics variant from the protocol
+version using the stored function. Note that the semantics variant depends on the language version
+too, but the latter is known statically (because each language version has its own evaluation
+context), hence there's no reason to require it to be provided at runtime.
+
+The reason why we associate a 'DefaultMachineParameters' with a semantics variant rather than a
+protocol version are
+
+1. generally there are far more protocol versions than semantics variants supported by a specific
+   language version, so we save on pointless duplication of bundles of machine parameters
+2. builtins don't know anything about protocol versions, only semantics variants. It is therefore
+   more semantically precise to associate bundles of machine parameters with semantics variants than
+   with protocol versions
 -}
-newtype EvaluationContext = EvaluationContext
-    { machineParameters :: DefaultMachineParameters
+data EvaluationContext = EvaluationContext
+    { _evalCtxLedgerLang    :: PlutusLedgerLanguage
+      -- ^ Specifies what language versions the 'EvaluationContext' is for.
+    , _evalCtxToSemVar      :: MajorProtocolVersion -> BuiltinSemanticsVariant DefaultFun
+      -- ^ Specifies how to get a semantics variant given a 'MajorProtocolVersion'.
```

Comment:
Should that really be "Specifies how to get a semantics variant _for this ledger language_ given a 'MajorProtocolVersion"?

Once again I'm becoming confused by all of the versions.  The PLC specification says (p 26)

>  To handle [changes in builtin semantics]  we define a series of built-in semantics variants, which indicate which
behaviour should be used. A fix will typically be deployed by defining a new semantics variant, and then
using that variant for future ledger languages (but not existing ones, since this is usually a backwards-
incompatible change).

That suggests that the semantics variant depends only on the ledger language, but I think we're now assuming that it depends on both the LL and the PV.  We should be careful to specify exactly what's going on.

Reply:
> That suggests that the semantics variant depends only on the ledger language, but I think we're now assuming that it depends on both the LL and the PV. 

Yes, exactly.